gk/fib.c

```diff
@@ -0,0 +1,1337 @@
+			"gk: failed to send a `struct gk_cmd_entry` entry at %s\n",
+			__func__);
+	}
+}
+
+/*
+ * For removing FIB entries, once the field @ref_cnt
+ * of a FIB entry becomes 0, then it needs to notify the GK instances
+ * to flush the flows that have a reference to it. This can be
+ * implemented by mailbox ring.
+ */
+static int
+del_fib_entry_locked(const char *ip_prefix, struct gk_config *gk_conf)
+{
+	int i, ret = 0;
+	int ip_prefix_present;
+	int ip_prefix_len;
+	struct ipaddr ip_prefix_addr;
+	struct gk_fib *ip_prefix_fib;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	if (ip_prefix == NULL)
+		return -1;
+
+ 	ip_prefix_len = parse_ip_prefix(ip_prefix, &ip_prefix_addr);
+	if (ip_prefix_len < 0)
+		return -1;
+
+	if (ip_prefix_addr.proto == ETHER_TYPE_IPv4) {
+		uint32_t fib_id;
+
+		ip_prefix_present = rte_lpm_is_rule_present(
+			ltbl->lpm, ntohl(ip_prefix_addr.ip.v4.s_addr),
+			ip_prefix_len, &fib_id);
+		if (!ip_prefix_present) {
+			RTE_LOG(WARNING, GATEKEEPER,
+				"gk: delete an invalid IP prefix (%s)\n",
+				ip_prefix);
+			return -1;
+		}
+
+		ip_prefix_fib = &ltbl->fib_tbl[fib_id];
+	} else if (ip_prefix_addr.proto == ETHER_TYPE_IPv6) {
+		uint8_t fib_id;
+
+		ip_prefix_present = rte_lpm6_is_rule_present(
+			ltbl->lpm6, ip_prefix_addr.ip.v6.s6_addr,
+			ip_prefix_len, &fib_id);
+		if (!ip_prefix_present) {
+			RTE_LOG(WARNING, GATEKEEPER,
+				"gk: delete an invalid IP prefix (%s)\n",
+				ip_prefix);
+			return -1;
+		}
+
+		ip_prefix_fib = &ltbl->fib_tbl6[fib_id];
+	} else {
+		RTE_LOG(WARNING, GATEKEEPER,
+			"gk: delete an invalid IP prefix (%s)\n",
+			ip_prefix);
+		return -1;
+	}
+
+	if (ip_prefix_fib->ref_cnt != 1) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: delete a fib entry [ip prefix = %s] with ref_cnt = %d, and action = %d\n",
+			ip_prefix, ip_prefix_fib->ref_cnt,
+			ip_prefix_fib->action);
+		return -1;
+	}
+
+	if (ip_prefix_addr.proto == ETHER_TYPE_IPv4) {
+		ret = rte_lpm_delete(ltbl->lpm,
+			ntohl(ip_prefix_addr.ip.v4.s_addr), ip_prefix_len);
+		if (ret < 0)
+			return -1;
+	} else {
+		ret = rte_lpm6_delete(ltbl->lpm6,
+			ip_prefix_addr.ip.v6.s6_addr, ip_prefix_len);
+		if (ret < 0)
+			return -1;
+	}
+
+	switch (ip_prefix_fib->action) {
+	case GK_FWD_GRANTOR: {
+		struct ip_flow *flow = &ip_prefix_fib->u.grantor.flow;
+		struct ether_cache *eth_cache =
+			ip_prefix_fib->u.grantor.eth_cache;
+
+		/* Decrement the @ref_cnt of the @next_fib entry. */
+		ip_prefix_fib->u.grantor.next_fib->ref_cnt--;
+		eth_cache->ref_cnt--;
+
+		/*
+		 * The Grantor is a neighbor, and the prefix
+		 * is the Grantor. So, we need to release the @eth_cache
+		 * Ethernet header entry from the neighbor hash table.
+		 */
+		if (ip_prefix_fib->u.grantor.next_fib->action ==
+				GK_FWD_NEIGHBOR_BACK_NET) {
+			/* Find the FIB entry for the Grantor. */
+			struct gk_fib *neigh_fib;
+			struct neighbor_hash_table *neigh_ht;
+
+			if (flow->proto == ETHER_TYPE_IPv4 && flow->f.v4.dst ==
+					ip_prefix_addr.ip.v4.s_addr) {
```

Comment:
Just to make sure that we are going to delete the FIB entry for the specific Grantor. When one adds a FIB entry for a Grantor, then we will initialize the ``struct ip_flow flow;`` field of the FIB entry. If they don't match, then there must be some bugs here. I may need to add one last ``else`` statement to deal with the bug situation.